x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java

```diff
@@ -283,6 +284,34 @@ private void createReservedUser(String username, char[] passwordHash, RefreshPol
         });
     }
 
+    /**
+     * Asynchronous method to create a reserved user with the given password hash. This should fail when the document for the
+     * specified user already exists in the security index
+     */
+    public void createReservedUser(String username, char[] passwordHash, ActionListener<Void> listener) {
+        securityIndex.prepareIndexIfNeededThenExecute(listener::onFailure, () -> {
+            executeAsyncWithOrigin(
+                client.threadPool().getThreadContext(),
+                SECURITY_ORIGIN,
+                client.prepareIndex(SECURITY_MAIN_ALIAS)
+                    .setOpType(DocWriteRequest.OpType.CREATE)
+                    .setId(getIdForUser(RESERVED_USER_TYPE, username))
+                    .setSource(
+                        Fields.PASSWORD.getPreferredName(),
+                        String.valueOf(passwordHash),
+                        Fields.ENABLED.getPreferredName(),
+                        true,
+                        Fields.TYPE.getPreferredName(),
+                        RESERVED_USER_TYPE
+                    )
+                    .setRefreshPolicy(RefreshPolicy.IMMEDIATE)
+                    .request(),
+                listener.<IndexResponse>delegateFailure((l, indexResponse) -> l.onResponse(null)),
```

Comment:
Are we sure we never need to clear the realm cache at this point?

Given this is a public method, with no indication that it should only be called at node startup, I think we should clear the cache.
If a reserved user is created _after_ someone has tried to authenticate with it, a node might have a negative cached password.

Reply:
Good point